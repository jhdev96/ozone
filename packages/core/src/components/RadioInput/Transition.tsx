import React from 'react';
import styles from './index.module.scss';


export default function Transition(): JSX.Element 
{
  return (
    <>
      <div className={styles["ripple-outer-ring"]}></div>
      <div className={styles["ripple-inner"]}></div>
    </>
  )
}