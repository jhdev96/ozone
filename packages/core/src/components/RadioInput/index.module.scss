@import '../../scss/root';
@import '../../scss/mixins/radio/animations';


.radioWrapper {
  display: block;
  position: relative;
  margin: 0.50rem 0;
  max-width: 15px;
  max-height: 15px;
  font-size: 22px;
  cursor: pointer;
}

.input {
  position: absolute;
  height: 0;
  width: 0;
  opacity: 0;
  cursor: pointer;
}

.input:checked ~ .radio {
  border-color: var(--blue-core);
  transition: border-color 0.6s;
}

.input:checked ~ .radio:after {
  display: block;
}

.input:disabled ~ .radio {
  background-color: rgba(0, 0, 0, 0.043);
  border: 2px solid #0000001d;

  .ripple-inner {
    background-color: #0000001d;
  }
}

.input:disabled ~ .radio:after {
  border-color: var(--gray-4);
}

.radio {
  position: absolute;
  width: 14px;
  height: 14px;
  border-radius: 50%;
  border: 2px solid var(--gray-3);
}

.text {
  vertical-align: middle;
  margin-left: 25px;
  color: var(--black-core);
}

.radio:after {
  content: "";
  position: absolute;
  display: none;
}

// Transition
.ripple-outer-ring {
  margin: auto;
  width: 3.5px;
  height: 3.5px;
  border: 1px solid var(--blue-core);
  border-radius: 50%;
  animation: ripple 0.4s ease-out forwards,
             fade 0.4s ease-out forwards;
}

.ripple-inner {
  position: absolute;
  top: 3px;
  left: 3px;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  opacity: 0;
  background-color: var(--blue-core);
  animation: fadeIn 0.4s ease-out 0.2s forwards;
}

// hasError prop
.error { 
  border-color: var(--red-core) !important;
  color: var(--red-core) !important; 

  .radio { border-color: var(--red-1); }

  .ripple-outer-ring {
    border: 1px solid var(--red-core);
  }

  .ripple-inner {
    background-color: var(--red-core);
  }

  ~ .text { color: var(--red-core) !important; }
}

// isDisabled prop
.disabled { color: var(--black-2) !important; }

// theme prop
.dark {
  .input:disabled ~ .radio {
    background-color: rgba(0, 0, 0, 0.043);
    border: 2px solid var(--black-2);
  
    .ripple-inner {
      background-color: var(--black-2);
    }
  }
  
  .text { color: var(--gray-core); }
}