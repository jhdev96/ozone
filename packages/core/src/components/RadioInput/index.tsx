import React, { useState, useEffect } from 'react';
import { ThemeType } from '../../utils/theme';
import classNames from '../../utils/classNames';
import Transition from './Transition';
import styles from './index.module.scss';


interface RadioInputProps {
  /** 
   * Control the lightness of the label text and radio border. 
   */
  theme?: ThemeType;
  /** 
   * Indicates whether the box is checked. 
   */
  isChecked?: boolean;
  /** 
   * Adds a 'required' HTML attribute if true. 
   */
  isRequired?: boolean;
  /** 
   * Adds a 'disabled' HTML attribute if true. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates visusally that there has been an error. 
   */
  hasError?: boolean;
  /** 
   * Gives a name to the checkbox for tracking input. 
   */
  name?: string;
  /** 
   * Callback to be invoked when a change is detected. 
   */
  onChange?: (checked: boolean, selected?: string) => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * content to be rendered beside the box. 
   */ 
  children?: React.ReactNode;
}

const RadioInput = React.forwardRef<HTMLInputElement, RadioInputProps>(
  function RadioInput({ 
    theme       = 'light',
    isChecked   = false, 
    isRequired  = true, 
    isDisabled  = false, 
    hasError    = false,
    onChange    = () => {},
    name,
    dataTestId,
    children}: RadioInputProps,
    ref: React.ForwardedRef<HTMLInputElement>): JSX.Element
  {
    const [checked, setChecked] = useState(isChecked);

    useEffect(() => {
      if (isChecked) setChecked(true);
      else setChecked(false);
    }, [isChecked]);

    const handleChange = () => {
      setChecked(!checked);
      onChange(checked, name);
    }

    const transition = checked && <Transition />;

    return (
      <label 
        className={styles[theme]} 
        data-testid={dataTestId}
      >
        <div 
          className={classNames(
            styles.radioWrapper, 
            hasError && styles.error
          )}>
          <input
            className={styles.input} 
            checked={checked}
            name={name}
            type="radio"
            required={isRequired}
            disabled={isDisabled}
            onChange={handleChange}
            ref={ref}
          />
          <span className={styles.radio}>
            {transition}
          </span>
        </div>
        <span 
          className={classNames(
            styles.text, 
            isDisabled && styles.disabled
          )}
        >
          {children}
        </span>
      </label>
    )
  }
)

export default RadioInput;

RadioInput.displayName = 'RadioInput';