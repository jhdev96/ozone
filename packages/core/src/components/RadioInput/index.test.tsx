import React from "react";
import { 
  render, 
  screen, 
  fireEvent
} from "@testing-library/react"; 
import RadioInput from ".";


describe('<RadioInput />', () => {
  it('renders a radio input with text content', () => {
    render(
      <RadioInput dataTestId="test">
        text
      </RadioInput>
    );

    expect(screen.getByTestId("test"))
      .toHaveTextContent("text");
  });

  it('renders an unchecked radio input by default', () => {
    render(
      <RadioInput dataTestId="test">
        text
      </RadioInput>
    );

    const radio = document.querySelector('.radio');
    expect(radio).toBeEmptyDOMElement();
  });

  it('sets the `required` attribute to true by default', () => {
    render(
      <RadioInput>text</RadioInput>
    );

    const input = document.querySelector('.input') as HTMLInputElement;
    expect(input.required).toBe(true);
  });

  it('sets the `name` attribute when specified', () => {
    render(
      <RadioInput name="hello">text</RadioInput>
    );

    const input = document.querySelector('.input') as HTMLInputElement;
    expect(input.name).toBe('hello');
  });

  it('renders a disabled radio input', () => {
    render(
      <RadioInput isDisabled>
        text
      </RadioInput>
    );

    const input = document.querySelector('.input') as HTMLInputElement;
    expect(input.disabled).toBe(true);
  });

  it('checks the radio input when the component is clicked', () => {
    render(
      <RadioInput dataTestId="test">
        text
      </RadioInput>
    );

    const input = document.querySelector('.input') as HTMLInputElement;

    expect(input.checked).toBe(false);

    fireEvent.click(input);

    expect(input.checked).toBe(true);
  });

  it('renders a transition when the radio input is checked', () => {
    render(
      <RadioInput dataTestId="test">
        text
      </RadioInput>
    );

    const transition = document.querySelector('.radio');
    fireEvent.click(screen.getByTestId('test'));

    expect(transition).not.toBeEmptyDOMElement();
  });

  it('calls `onChange` when the input changes', () => {
    const onChange = jest.fn();

    render(
      <RadioInput 
        onChange={onChange} 
        dataTestId="test"
      >
        text
      </RadioInput>
    );

    fireEvent.click(screen.getByTestId('test'));

    expect(onChange).toBeCalled;
  });

  it('renders a radio input with an error', () => {
    render(
      <RadioInput dataTestId="test" hasError>
        text
      </RadioInput>
    );

    expect(screen.getByTestId('test'))
      .toMatchSnapshot();
  });
});