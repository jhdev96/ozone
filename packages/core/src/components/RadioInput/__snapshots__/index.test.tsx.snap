// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioInput /> renders a radio input with an error 1`] = `
<label
  class="light"
  data-testid="test"
>
  <div
    class="radioWrapper error"
  >
    <input
      class="input"
      required=""
      type="radio"
    />
    <span
      class="radio"
    />
  </div>
  <span
    class="text"
  >
    text
  </span>
</label>
`;
